### DISCUSSION TITLE
List query params, Pydantic, Field, Depends and Annotated - confused...

### DISCUSSION THREAD

**Discussion by wabiloo** [fastapi/fastapi#11333] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated, List

import uvicorn
from fastapi import Depends, FastAPI, Query
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

app = FastAPI()


class WorkingParams(BaseModel):
    crea: List[str] = Field(Query(..., description="Creative URLs"))
    click: List[str] = Field(Query([], description="Click URLs"))
    ver: float = Field(Query(3.0, description="VAST version"))


class NotWorkingParams(BaseModel):
    crea: Annotated[List[str], Field(Query(description="Creative URLs"))]
    click: Annotated[List[str], Field(Query(description="Click URLs"))] = []
    ver: Annotated[float, Field(Query(description="VAST version"))] = 3.0


@app.get("/working")
def read_root(params: WorkingParams = Depends(WorkingParams)):
    return JSONResponse(params.model_dump())


@app.get("/not-working")
def read_root(params: NotWorkingParams = Depends(NotWorkingParams)):
    return JSONResponse(params.model_dump())


if __name__ == "__main__":
    uvicorn.run("main:app", host="127.0.0.1", reload=True)
```

### Description

I am attempting to use Pydantic models for query params, including params that are lists. Some are mandatory, some not.
I am also trying to use the "modern" way of defining the fields of the model, with Annotated.

However I noticed that if I do that, the optional List[str] params are not being returned.

http://127.0.0.1:8000/working?crea=crea1&click=click1&crea=crea2&ver=4
returns 
```
{
    "crea": [
        "crea1",
        "crea2"
    ],
    "click": [
        "click1"
    ],
    "ver": 4.0
}
```
But http://127.0.0.1:8000/not-working?crea=crea1&click=click1&crea=crea2&ver=4
returns
```
{
    "crea": [
        "crea1",
        "crea2"
    ],
    "click": [],
    "ver": 4.0
}
```

The `click` parameter has the wrong default value

### Operating System

macOS

### FastAPI Version

0.110.0

### Pydantic Version

2.6.4

### Python Version

3.11.7

**Comment by YuriiMotov** [comment, marked as answer]
Don't use Pydantic models for query params because it's not officially supported yet.
Take a look at the answer of Sebastián Ramírez to this comment: https://github.com/tiangolo/fastapi/discussions/9709#discussioncomment-6532693, also [this](https://github.com/tiangolo/fastapi/discussions/9709#discussioncomment-6624315)

You can use dataclasses or usual classes for Query params

**Comment by JoshYuJump** [comment]
@wabiloo The alternative is to downgrade to Pydantic v1.

**Reply by wabiloo** [reply]
That's a shame. I have an application that has both a CLI mode and a server with API mode (with FastAPI).

Using Pydantic models would allow me to easily define a single set of function parameters that can be used to define both (with argparse-pydantic)
